src/main/java/hudson/plugins/gradle/injection/MavenOptsSetter.java

```diff
@@ -17,54 +17,81 @@ class MavenOptsSetter {
     private static final String MAVEN_OPTS_VAR = "MAVEN_OPTS";
 
     private final Set<String> keys;
+    private final Set<String> requiredKeys;
 
-    public MavenOptsSetter(String... keys) {
-        this.keys = ImmutableSet.copyOf(keys);
+    public MavenOptsSetter(SystemProperty.Key... keys) {
+        this.keys =
+            Arrays.stream(keys)
+                .map(k -> k.name)
+                .collect(
+                    Collectors.collectingAndThen(Collectors.toSet(), Collections::unmodifiableSet));
+        this.requiredKeys =
+            Arrays.stream(keys)
+                .filter(k -> k.required)
+                .map(k -> k.name)
+                .collect(
+                    Collectors.collectingAndThen(Collectors.toSet(), Collections::unmodifiableSet));
     }
 
-    void appendIfMissing(Node node, List<SystemProperty> systemProperties) throws IOException, InterruptedException {
-        String geMavenOpts =
+    void append(Node node, List<SystemProperty> systemProperties) throws IOException, InterruptedException {
+        String additionalProperties =
             systemProperties
                 .stream()
                 .map(SystemProperty::asString)
                 .collect(Collectors.joining(SPACE));
 
-        String mavenOpts = removeSystemProperties(getMavenOpts(node)) + SPACE + geMavenOpts;
+        String mavenOpts =
+            filtered(getCurrentMavenOpts(node))
+                .map(current -> String.join(SPACE, current, additionalProperties))
+                .orElse(additionalProperties);
+
         EnvUtil.setEnvVar(node, MAVEN_OPTS_VAR, mavenOpts);
     }
 
-    void remove(Node node) throws IOException, InterruptedException {
-        String mavenOpts = removeSystemProperties(getMavenOpts(node));
+    void removeIfNeeded(Node node) throws IOException, InterruptedException {
+        String currentMavenOpts = getCurrentMavenOpts(node);
+        if (currentMavenOpts == null || currentMavenOpts.isEmpty()) {
+            return;
+        }
+
+        boolean hasAllRequiredKeys = requiredKeys.stream().allMatch(currentMavenOpts::contains);
+        if (!hasAllRequiredKeys) {
+            return;
+        }
+
+        String mavenOpts = filtered(currentMavenOpts).orElse(null);
         EnvUtil.setEnvVar(node, MAVEN_OPTS_VAR, mavenOpts);
     }
 
-    private String getMavenOpts(Node node) throws IOException, InterruptedException {
-        EnvVars nodeEnvVars = EnvVars.getRemote(node.getChannel());
-        return nodeEnvVars.get(MAVEN_OPTS_VAR);
+    @Nullable
+    private static String getCurrentMavenOpts(Node node) {
+        return EnvUtil.getEnv(node, MAVEN_OPTS_VAR);
```

Comment:
What about changing `EnvUtil.getEnv` to return an `Optional` directly ?

Reply:
It's possible. But I'm not sure about the value though. This is a private method with a very limited scope. It's quite easy to ensure that `null` value is properly handled. Not all usages will really benefit from an Optional here. See `removeIfNeeded` method for example, where we would need to resolve optional straightaway to be able to perform a check.